src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkWasmModule.java

```diff
@@ -0,0 +1,278 @@
+import java.time.Duration;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import net.starlark.java.annot.StarlarkBuiltin;
+import net.starlark.java.annot.StarlarkMethod;
+import net.starlark.java.eval.EvalException;
+import net.starlark.java.eval.Printer;
+import net.starlark.java.eval.Starlark;
+import net.starlark.java.eval.StarlarkValue;
+
+@Immutable
+@StarlarkBuiltin(
+    name = "wasm_module",
+    category = DocCategory.BUILTIN,
+    doc = "A WebAssembly module loaded by <code>repository_ctx.load_wasm()</code>.")
+final class StarlarkWasmModule implements StarlarkValue {
+  private final StarlarkPath path;
+  private final Object origPath;
+  private final WasmModule wasmModule;
+  private final String allocateFn;
+
+  public StarlarkWasmModule(
+      StarlarkPath path,
+      Object origPath,
+      byte[] moduleContent,
+      String allocateFn)
+      throws EvalException {
+    WasmModule wasmModule;
+    Profiler prof = Profiler.instance();
+    try (SilentCloseable c1 = prof.profile(WASM_LOAD, () -> "load " + path.toString())) {
+      try (SilentCloseable c2 = prof.profile(WASM_LOAD, "parse")) {
+        try {
+          wasmModule = com.dylibso.chicory.wasm.Parser.parse(moduleContent);
+        } catch (ChicoryException e) {
+          throw new EvalException(e);
+        }
+      }
+      validateModule(wasmModule, allocateFn);
+    }
+
+    this.path = path;
+    this.origPath = origPath;
+    this.wasmModule = wasmModule;
+    this.allocateFn = allocateFn;
+  }
+
+  @Override
+  public boolean isImmutable() {
+    return true;
+  }
+
+  @Override
+  public void repr(Printer printer) {
+    printer.append("<wasm_module path=");
+    printer.repr(origPath);
+    printer.append(">");
+  }
+
+  public StarlarkPath getPath() {
+    return path;
+  }
+
+  @StarlarkMethod(
+      name = "path",
+      structField = true,
+      doc = "The path this WebAssembly module was loaded from.")
+  public Object getOrigPath() {
+    return origPath;
+  }
+
+  public StarlarkWasmExecutionResult execute(
+      String functionName,
+      byte[] input,
+      Duration timeout,
+      long memLimitBytes)
+      throws EvalException, InterruptedException {
+    Profiler prof = Profiler.instance();
+    try (SilentCloseable c = prof.profile(WASM_EXEC, () -> "execute " + functionName)) {
+      var memLimits = getMemLimits(memLimitBytes);
+      // Perform initialization and execution in a separate thread so it can be interrupted
+      // in case of timeout.
+      var wasmThreadFactory =
+          Thread.ofPlatform().name(Thread.currentThread().getName() + "_wasm").factory();
+      StarlarkWasmExecutionResult result;
+      String errMessage;
+      try (var executor = Executors.newSingleThreadExecutor(wasmThreadFactory)) {
+        return executor.invokeAny(
```

Comment:
any reason to use `invokeAny` over just `submit(...).get(...)`? There's only ever one task here.

Reply:
The original version of this code used `submit(); get()`, and it was changed to the current syntax by reviewer request.

I have no strong preference about how to use this API, and would be fine with changing the code back if that's the Bazel team's preference.

Reply:
@Wyverald I suggested this and am okay with it being changed back if you prefer that.

To me, Futures are a low-level primitive that's best avoided if a more high-level, structured alternative exists. `invokeAny`'s behavior regarding cancellation is pretty much determined by its type signature (it returns a bare value), which helps readability and avoids misuse. For example, a Future would need to be cancelled manually when `get` times out.

Reply:
ah ok, that's interesting. I wasn't aware of this idiom. I'm ok with leaving it as is, although note that you also don't need to cancel the future manually; a shutdownNow() on the executor service would suffice.